### PR TITLE
Support token-based auth for cron-connector

### PR DESCRIPTION
**Pull request by welteki**
## Description

Add support for token based authentication with OpenFaaS IAM to the cron-connector chart.

When IAM is enabled a projected ServiceAccount token volume is mounted into the pod. The connector can exchange this token for an OpenFaaS token that is in turn used to authenticate with the gateway API.

The chart automatically creates the required `Policy`, `Role` and `ServiceAccount` to authenticate the connector with minimal privileges.

The configuration for IAM is included under the `iam` section in the values file:

```yaml
openfaasPro: true

iam:
  enabled: true
  systemIssuer:
    url: "https://gateway.openfaas.example.com"
  resource:
    - "dev:*"
```

## Why is this needed?

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

- Enables the connector to access the Gateway with minimal privilege
- Restrict the resources a connector deployment can operate on. Useful for multi-tenancy.

## Who is this for?

What company is this for? Are you listed in the [ADOPTERS.md](https://github.com/openfaas/faas/blob/master/ADOPTERS.md) file?

## How Has This Been Tested?

Verified the connector can be used with basic auth, token based authentication for IAM and with authentication disabled.

Verified the correct `Role`, `Policy` and `ServiceAccount` based on the release name are created.

values.yaml file;

```yaml
openfaasPro: true

basicAuth: false

iam:
  enabled: true
  resource:
    - "dev:*"
```

Run template and verify output:
```
helm template --namespace openfaas \
  cron-connector \
  ./chart/cron-connector \
  -f values.yaml
```
output:

```yaml
---
# Source: cron-connector/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: cron-connector
  namespace: "openfaas"
  labels:
    helm.sh/chart: cron-connector-0.6.4
    app.kubernetes.io/name: cron-connector
    app.kubernetes.io/instance: cron-connector
    app.kubernetes.io/component: cron-connector
    app.kubernetes.io/part-of: openfaas
    app.kubernetes.io/managed-by: Helm
---
# Source: cron-connector/templates/role.yaml
apiVersion: iam.openfaas.com/v1
kind: Role
metadata:
  name: cron-connector
  namespace: openfaas
  labels:
    helm.sh/chart: cron-connector-0.6.4
    app.kubernetes.io/name: cron-connector
    app.kubernetes.io/instance: cron-connector
    app.kubernetes.io/component: cron-connector
    app.kubernetes.io/part-of: openfaas
    app.kubernetes.io/managed-by: Helm
spec:
  policy:
  - cron-connector
  principal:
    jwt:sub:
      - "system:serviceaccount:openfaas:cron-connector"
  condition:
    StringEqual:
      jwt:iss:
        - https://kubernetes.default.svc.cluster.local
---
# Source: cron-connector/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    # Original Helm labels v
    app: cron-connector
    component: cron-connector
    chart: cron-connector-0.6.4
    heritage: Helm
    release: cron-connector
    helm.sh/chart: cron-connector-0.6.4
    app.kubernetes.io/name: cron-connector
    app.kubernetes.io/instance: cron-connector
    app.kubernetes.io/component: cron-connector
    app.kubernetes.io/part-of: openfaas
    app.kubernetes.io/managed-by: Helm
  name: cron-connector
  namespace: "openfaas"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: cron-connector
      name: cron-connector
      component: cron-connector
  template:
    metadata:
      labels:
        app: cron-connector
        name: cron-connector
        component: cron-connector
    spec:
      serviceAccountName: cron-connector
      volumes:
        - name: license
          secret:
            secretName: openfaas-license
        - name: openfaas-token
          projected:
            sources:
            - serviceAccountToken:
                path: openfaas-token
                expirationSeconds: 7200
                audience: https://gateway.example.com
      containers:
        - name: cron-connector
          image: ghcr.io/openfaasltd/cron-connector:0.2.0
          command:
            - "/usr/bin/cron-connector"
            - "-license-file=/var/secrets/license/license"
          env:
            - name: gateway_url
              value: "http://gateway.openfaas:8080"
            - name: asynchronous_invocation
              value: "false"
            - name: content_type
              value: "text/plain"
            - name: rebuild_interval
              value: "30s"
            - name: "debug"
              value: "false"
            - name: "log_encoding"
              value: "console"
            - name: system_issuer
              value: https://gateway.example.com
          resources:
          requests:
            cpu: 100m
            memory: 64Mi
          volumeMounts:
            - name: license
              readOnly: true
              mountPath: "/var/secrets/license"
            - name: openfaas-token
              readOnly: true
              mountPath: /var/secrets/tokens
---
# Source: cron-connector/templates/policy.yaml
apiVersion: iam.openfaas.com/v1
kind: Policy
metadata:
  name: cron-connector
  namespace: openfaas
  labels:
    helm.sh/chart: cron-connector-0.6.4
    app.kubernetes.io/name: cron-connector
    app.kubernetes.io/instance: cron-connector
    app.kubernetes.io/component: cron-connector
    app.kubernetes.io/part-of: openfaas
    app.kubernetes.io/managed-by: Helm
spec:
  statement:
  - sid: 1-list
    action:
    - Function:List
    - Namespace:List
    effect: Allow
    resource:
      - dev:*
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by welteki** [comment]
This needs a new release of the cron-connector so I can remove the draft status.

File: chart/cron-connector/templates/policy.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+metadata:
+  name: {{ template "connector.fullname" . }}
+  namespace: openfaas
+  labels:
+    {{- include "connector.labels" . | nindent 4 }}
+spec:
+  statement:
+  - sid: 1-fn-r
+    action:
+    - Function:Read
```
Comment (alexellis) [inline chart/cron-connector/templates/policy.yaml:15]: This should be Function:List, perhaps also Function:Get?

Comment (welteki) [reply chart/cron-connector/templates/policy.yaml]: The required actions are:
- Function:List
- Namespace:List

This has been updated and tested.

**Review by alexellis — APPROVED** [review]
Approved

**Pull request merged**